### PR TITLE
Allow specifying docker file for tye run

### PR DESCRIPTION
**Pull request by jkotalik**
Fixes https://github.com/dotnet/tye/issues/146

I don't know if well get this into the next release (I'm reluctant to just because I haven't finished the tye deploy story). There are also some paper cuts that are going to make it hard for adoption, including not automatically allocating ports for docker projects (looking into this).

### CODE REVIEW

**Comment by jkotalik** [comment]
Something is going on with the skip conditions for Docker on Windows. Last four builds have failed due to tests not being skipped. I'll look into it.

File: src/Microsoft.Tye.Hosting/DockerRunner.cs [hunk 1]
```diff
@@ -195,6 +195,20 @@ private async Task StartContainerAsync(Application application, Service service,
             var volumes = "";
             var workingDirectory = docker.WorkingDirectory != null ? $"-w {docker.WorkingDirectory}" : "";
             var hostname = "host.docker.internal";
+            var dockerImage = docker.Image ?? service.Description.Name;
+
+            if (docker.DockerFile != null)
+            {
+                var exitCode = await ProcessUtil.RunAsync(
+                    $"docker",
+                    $"build \"{docker.DockerFile.DirectoryName}\" -t {dockerImage} -f \"{docker.DockerFile}\"",
```
Comment (davidfowl) [inline src/Microsoft.Tye.Hosting/DockerRunner.cs:204]: We need to support a different build context as well.

Comment (jkotalik) [reply src/Microsoft.Tye.Hosting/DockerRunner.cs:204]: How would you recommend exposing that? In the tye.yaml as another part of a service?

File: test/E2ETest/testassets/projects/dockerfile/tye.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# tye application configuration file
+# read all about it at https://github.com/dotnet/tye
+name: frontend-backend
+services:
+- name: backend
+  dockerFile: backend/Dockerfile
+  bindings:
+    - port: 8002
```
Comment (davidfowl) [inline test/E2ETest/testassets/projects/dockerfile/tye.yaml:8]: Random port

Comment (davidfowl) [reply test/E2ETest/testassets/projects/dockerfile/tye.yaml:8]: You should be able to set containerPort to 80 (but I think it'll default for http) and set nothing else

Comment (jkotalik) [reply test/E2ETest/testassets/projects/dockerfile/tye.yaml:8]: That needs to be fixed. Right now, a random port isn't specified (and I don't think one is specified for any non-project, but need to confirm that).

Comment (davidfowl) [reply test/E2ETest/testassets/projects/dockerfile/tye.yaml:8]: Not sure what you mean. If you set containerPort a random port will get generated

File: src/Microsoft.Tye.Hosting/DockerRunner.cs (hunk 1 above)
Comment (davidfowl) [reply src/Microsoft.Tye.Hosting/DockerRunner.cs:204]: Yep

**Comment by rynowak** [comment]
> I don't know if well get this into the next release (I'm reluctant to just because I haven't finished the tye deploy story). There are also some paper cuts that are going to make it hard for adoption, including not automatically allocating ports for docker projects (looking into this).

The port issue needs to be fixed 👍 

I don't really mind if this goes in without deploy support at first. As long as we can document what's working, and it's useful to somebody there's no reason to delay it.

**Comment by jkotalik** [comment]
> The port issue needs to be fixed 👍

So apparently, I didn't know this, but you need to specify the containerPort. Random ports do seem to work if you specify that.

**Comment by davidfowl** [comment]
@amarie401 time to try that first scenario again

**Pull request merged**